src/OrchardCore.Modules/OrchardCore.Queries/Permissions.cs

```diff
@@ -31,7 +31,9 @@ public async Task<IEnumerable<Permission>> GetPermissionsAsync()
             ExecuteApiAll,
         };
 
-        foreach (var query in await _queryManager.ListQueriesAsync())
+        var queries = await _queryManager.ListQueriesAsync(true);
```

Comment:
It may be good to cache the permissions too, because those may be generated during each request. We can reuse the same concept as with the GraphQL types.

Reply:
@gvkries I don't think this is needed. The `GetPermissionsAsync` is not called on every request. I think it is only called when we list permission when editing a role and when we add new admin menu item.

Reply:
@Piedone are you saying here "Yup" as it we should cache it?

Reply:
The permissions are always build on every request, e.g. if you have an admin menu with permissions assigned. I had the same conversation with @Piedone while working on the media folder permissions.

Reply:
`GetPermissionsAsync()` is only called when listing permissions for a role, i.e. on the role edit pages by default. This isn't called for every page view, since it's not called for `IAuthorizationService.AuthorizeAsync()`. So, it doesn't need to be cached.

Reply:
@Piedone I'll look at it. This sounds like an issue in the AdminMenu that should be addressed not here. The AdminMenu may need to handle the cashing for all permissions.

Reply:
> @Piedone I'll look at it. This sounds like an issue in the AdminMenu that should be addressed not here. The AdminMenu may need to handle the cashing for all permissions.

I don't think that is a solution, as the admin menu should not know about the permissions providers. Invalidating a cache is too specific for each case.

Reply:
@gvkries the following code is called on every request as you stated. I was not aware of this.

https://github.com/OrchardCMS/OrchardCore/blob/6312841ef0893e1585e93e325289794de4aada94/src/OrchardCore.Modules/OrchardCore.AdminMenu/Services/AdminMenuPermissionService.cs#L28-L33

Why can't we have `PermissionsManager` that would handler getting the permissions and also caching them in memory which will solve this problem and any other problem related to caching permissions? 

Here is another existing instance where we query the roles on every request. I am betting will find other places as well. So caching this using `PermissionsManager` may be a safe approach instead of having to deal with caching every thing all over the place.

https://github.com/OrchardCMS/OrchardCore/blob/6312841ef0893e1585e93e325289794de4aada94/src/OrchardCore.Modules/OrchardCore.Roles/Permissions.cs#L31

Reply:
We could add caching on a higher level, though then a (distributed?) signal is needed to invalidate it from all over the place.

Reply:
If we cache the permissions in a singleton object, we won't have to worry about invalidating since these objects will be reinitialized when the tenant is reloaded.

Reply:
A singleton is not enough, we must invalidate the cache in a variety of reasons. E.g. in this case for adding or deleting a query. The tenant is not reloaded under these circumstances.

Reply:
Yep. We can cover this in a separate PR though.

Reply:
This is no longer an issue for this PR since I moved into using a new Document to manage all these queries. But we should think about a way to generate higher level cache or address any existing issue with permission like the Roles permission not cached.

Reply:
I opened an issue to track this: https://github.com/OrchardCMS/OrchardCore/issues/16416.